analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarAnalysisContextBase.cs

```diff
@@ -18,30 +18,47 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Collections.Concurrent;
 using System.IO;
 using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Text;
 using static SonarAnalyzer.Helpers.DiagnosticDescriptorFactory;
 
 namespace SonarAnalyzer.AnalysisContext;
 
 public class SonarAnalysisContextBase
 {
+    protected static readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<string, bool>> IncludedExcludedFilesCache = new();
     protected static readonly ConditionalWeakTable<Compilation, ImmutableHashSet<string>> UnchangedFilesCache = new();
     protected static readonly SourceTextValueProvider<ProjectConfigReader> ProjectConfigProvider = new(x => new ProjectConfigReader(x));
     private static readonly Lazy<SourceTextValueProvider<bool>> ShouldAnalyzeGeneratedCS = new(() => CreateAnalyzeGeneratedProvider(LanguageNames.CSharp));
     private static readonly Lazy<SourceTextValueProvider<bool>> ShouldAnalyzeGeneratedVB = new(() => CreateAnalyzeGeneratedProvider(LanguageNames.VisualBasic));
+    private static readonly Lazy<SourceTextValueProvider<string[]>> FileExclusions = new(() => CreateFileExclusions());
+    private static readonly Lazy<SourceTextValueProvider<string[]>> FileInclusions = new(() => CreateFileInclusions());
 
     protected SonarAnalysisContextBase() { }
 
     protected static SourceTextValueProvider<bool> ShouldAnalyzeGeneratedProvider(string language) =>
         language == LanguageNames.CSharp ? ShouldAnalyzeGeneratedCS.Value : ShouldAnalyzeGeneratedVB.Value;
 
+    protected static SourceTextValueProvider<string[]> RetrieveExcludedFiles() => FileExclusions.Value;
+
+    protected static SourceTextValueProvider<string[]> RetrieveIncludedFiles() => FileInclusions.Value;
+
     private static SourceTextValueProvider<bool> CreateAnalyzeGeneratedProvider(string language) =>
         new(x => PropertiesHelper.ReadAnalyzeGeneratedCodeProperty(PropertiesHelper.ParseXmlSettings(x), language));
+
+    private static SourceTextValueProvider<string[]> CreateFileExclusions() =>
+        new(x => PropertiesHelper.ReadSourceFileExclusionsProperty(PropertiesHelper.ParseXmlSettings(x)));
```

Comment:
With this implementation we are parsing the SonarLint.xml file 4 times (2 for CreateAnalyzeGeneratedProvider, one for each language, 1 for CreateFileExclusions and 1 for CreateFileInclusions). This can be further improved on the implementation sprint. It shouldn't be a concern for the performance evaluation since the computational cost of these four operations is minimal.

As a result of this decision we have 4 failed UTs (the scope of this PR is not to merge it to master so we are ok with that).
- ShouldAnalyzeTree_CorrectSettingUsed: the ToString() call should be 2 (cs/vb), now they are 4, one for each parse attempt.
- ShouldAnalyzeTree_GeneratedFile_InvalidSonarLintXml: the ToString() call should be 1 (cs), now they are 3.
- ShouldAnalyzeTree_GeneratedFile_ShouldAnalyzeGeneratedProvider_IsCached: expected 3 invocations on the mock but was 9 (3x3)